File: 2023/haskell/day03b/Main.hs
module Main ( main ) where

import Data.Char
import qualified Data.Map as Map
import Data.Maybe
import System.Environment
import System.Exit
import System.IO

data Position = Position { r :: Int, c :: Int }
  deriving (Eq, Ord, Show)

buildNumbers :: String -> Map.Map Position String
buildNumbers contents =
  let (_, numberLocs) = foldl
                          (\(row, numberLocs) line ->
                              let checkLine (col, scanningNumber, number, currentPos, numLocs) ch =
                                    if scanningNumber
                                      then
                                        if (isDigit ch)
                                          then
                                            if (col + 1) == (length line)
                                              then (col + 1, False, "", Position (-1) (-1), Map.insert currentPos (number ++ [ch]) numLocs)
                                              else (col + 1, True, number ++ [ch], currentPos, numLocs)
                                          else (col + 1, False, "", Position (-1) (-1), Map.insert currentPos number numLocs)
                                      else
                                        if (isDigit ch)
                                          then (col + 1, True, [ch], Position row col, numLocs)                                    
                                          else (col + 1, False, number, currentPos, numLocs)
                                  (_, _, _, _, numberLocs') = foldl checkLine (0, False, "", Position (-1) (-1), numberLocs) line
                              in (row + 1, numberLocs')
                          )
                          (0, Map.empty)
                          (lines contents)
  in numberLocs

buildGears :: String -> Map.Map Position Char
buildGears contents =
  let (_, gearLocs) = foldl
                        (\(row, gearLocs) line ->
                            let checkLine (col, grLocs) ch =
                                  if ch == '*'
                                    then (col + 1, Map.insert (Position row col) ch grLocs)
                                    else (col + 1, grLocs)
                                (_, gearLocs') = foldl checkLine (0, gearLocs) line
                            in (row + 1, gearLocs')
                        )
                        (0, Map.empty)
                        (lines contents)
  in gearLocs

checkGears :: Map.Map Position String -> Map.Map Position Char -> Int
checkGears numberLocs gearLocs =
  foldl
    (\result gearLoc ->
      let checkAdjacents adjacentsList numberLoc =
            let number = fromJust $ Map.lookup numberLoc numberLocs
                numberRow = r numberLoc
                numberColFirst = c numberLoc
                numberColLast = (c numberLoc) + (length number) - 1
                neighbors = [ Position (-1) (-1), Position (-1) 0, Position (-1) 1,
                              Position 0 (-1), Position 0 1,
                              Position 1 (-1), Position 1 0, Position 1 1 ]
                checkNeighbor found neighbor =
                  let adjacentPos = Position ((r gearLoc) + (r neighbor)) ((c gearLoc) + (c neighbor))
                  in found || (((r adjacentPos) == numberRow) &&
                                ((c adjacentPos) >= numberColFirst) &&
                                ((c adjacentPos) <= numberColLast))
                foundAdjacent = foldl checkNeighbor False neighbors
            in
              if foundAdjacent
                then (adjacentsList ++ [(read number)])
                else adjacentsList
          adjacents = foldl checkAdjacents [] (Map.keys numberLocs)
      in
        if (length adjacents) == 2
          then (result + (product adjacents))
          else result
    )
    0
    (Map.keys gearLocs)

process :: String -> Int
process contents =
  let numberLocs = buildNumbers contents
      gearLocs = buildGears contents
  in checkGears numberLocs gearLocs

usage :: IO a
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitWith (ExitFailure 1)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [] -> usage
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn $ "result = " ++ show result
    _ -> usage

File: 2023/java/day03b/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Main {
    static class Position {
        public int row;
        public int col;

        Position(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    static void usage() {
        System.out.println("usage: java Main <file>");
        System.exit(1);
    }

    static HashMap<Position, String> buildNumbers(String filename) {
        HashMap<Position, String> numberLocs = new HashMap<Position, String>();
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));
            String line;
            int row = 0;
            boolean scanningNumber = false;
            StringBuffer number = new StringBuffer();
            Position currentPos = new Position(-1, -1);
            while ( (line = reader.readLine()) != null ) {
                for ( int col = 0; col < line.length(); col++ ) {
                    char ch = line.charAt(col);
                    if ( scanningNumber ) {
                        if ( Character.isDigit(ch) ) {
                            number.append(ch);
                        } else {
                            numberLocs.put(currentPos, number.toString());
                            scanningNumber = false;
                            number = new StringBuffer();
                        }
                    } else {
                        if ( Character.isDigit(ch) ) {
                            number.append(ch);
                            currentPos = new Position(row, col);
                            scanningNumber = true;
                        }
                    }
                }
                if ( scanningNumber ) {
                    numberLocs.put(currentPos, number.toString());
                    scanningNumber = false;
                    number = new StringBuffer();
                }
                row++;
            }
            reader.close();
        } catch ( IOException e ) {
            e.printStackTrace();
        }
        return numberLocs;
    }

    static HashMap<Position, Character> buildParts(String filename) {
        HashMap<Position, Character> partLocs = new HashMap<Position, Character>();
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));
            String line;
            int row = 0;
            while ( (line = reader.readLine()) != null ) {
                for ( int col = 0; col < line.length(); col++ ) {
                    char ch = line.charAt(col);
                    if ( ch == '*' ) {
                        partLocs.put(new Position(row, col), ch);
                    }
                }
                row++;
            }
            reader.close();
        } catch ( IOException e ) {
            e.printStackTrace();
        }
        return partLocs;
    }

    static int checkParts(HashMap<Position, String> numLocs, HashMap<Position, Character> partLocs) {
        int result = 0;
        ArrayList<Position> neighbors = new ArrayList<Position>();
        for ( int i = -1; i <= 1; i++ ) {
            for ( int j = -1; j <= 1; j++ ) {
                if ( !( i == j && i == 0) ) {
                    neighbors.add(new Position(i, j));
                }
            }
        }
        for ( Map.Entry<Position, Character> partLocEntry : partLocs.entrySet() ) {
            Position partLoc = partLocEntry.getKey();
            int adjacentCount = 0;
            int prod = 1;
            for ( Map.Entry<Position, String> numLocEntry : numLocs.entrySet() ) {
                Position numLoc = numLocEntry.getKey();
                String number = numLocEntry.getValue();
                int numRow = numLoc.row;
                int numColFirst = numLoc.col;
                int numColLast = numLoc.col + number.length();
                boolean found = false;
                for ( Position neighbor : neighbors ) {
                    Position adjacent = new Position(partLoc.row + neighbor.row, partLoc.col + neighbor.col);
                    for ( int numCol = numColFirst; numCol < numColLast; numCol++ ) {
                        if ( numRow == adjacent.row && numCol == adjacent.col ) {
                            adjacentCount++;
                            prod *= Integer.parseInt(number);
                            found = true;
                            break;
                        }
                    }
                    if ( found ) {
                        break;
                    }
                }
            }
            if ( adjacentCount == 2 ) {
                result += prod;
            }
        }
        return result;
    }

    static int process(String filename) {
        HashMap<Position, String> numLocs = buildNumbers(filename);
        HashMap<Position, Character> partLocs = buildParts(filename);
        return checkParts(numLocs, partLocs);
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + Integer.toString(result));
    }
}
